File: problems/0001-two-sum/java/Solution.java
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

// LeetCode 1. Two Sum
// Single-file program: run with `java -ea Solution.java` (assertions enabled).
public class Solution {
    static int[] twoSum(int[] nums, int target) {
        Map<Integer, Integer> seen = new HashMap<>(); // value -> index
        for (int i = 0; i < nums.length; i++) {
            Integer j = seen.get(target - nums[i]);
            if (j != null) {
                return new int[] {j, i};
            }
            seen.put(nums[i], i);
        }
        return new int[] {};
    }

    public static void main(String[] args) {
        assert Arrays.equals(twoSum(new int[] {2, 7, 11, 15}, 9), new int[] {0, 1});
        assert Arrays.equals(twoSum(new int[] {3, 2, 4}, 6), new int[] {1, 2});
        assert Arrays.equals(twoSum(new int[] {3, 3}, 6), new int[] {0, 1});
        System.out.println("All Two Sum tests passed.");
    }
}

File: problems/0001-two-sum/kotlin/Solution.kt
// LeetCode 1. Two Sum
// Run: kotlinc Solution.kt -include-runtime -d solution.jar && java -jar solution.jar

fun twoSum(nums: IntArray, target: Int): IntArray {
    val seen = HashMap<Int, Int>() // value -> index
    for ((i, n) in nums.withIndex()) {
        seen[target - n]?.let { return intArrayOf(it, i) }
        seen[n] = i
    }
    return intArrayOf()
}

fun main() {
    check(twoSum(intArrayOf(2, 7, 11, 15), 9).contentEquals(intArrayOf(0, 1)))
    check(twoSum(intArrayOf(3, 2, 4), 6).contentEquals(intArrayOf(1, 2)))
    check(twoSum(intArrayOf(3, 3), 6).contentEquals(intArrayOf(0, 1)))
    println("All Two Sum tests passed.")
}
